fix(core): fix ng generate @angular/core:output-migration

aparzi opened the pull request:
output-migration command not keep type if @Output declaring without initializer

## PR Checklist
Please check if your PR fulfills the following requirements:

- [X] The commit message follows our guidelines: https://github.com/angular/angular/blob/main/contributing-docs/commit-message-guidelines.md
- [X] Tests for the changes have been added (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

- [X] Bugfix


## What is the current behavior?

The output-migration command not keep type if @Output declaring without initializer


Issue Number: #59248 

## What is the new behavior?

The output-migration command keep type

## Does this PR introduce a breaking change?

- [ ] Yes
- [X] No

JeanMeche commented on packages/core/schematics/migrations/output-migration/output-replacements.ts:
```diff
@@ -28,10 +28,16 @@ export function calculateDeclarationReplacement(
+    payloadTypes = node.initializer.typeArguments;
+  }
+
+  if (!payloadTypes && node.type && ts.isTypeReferenceNode(node.type) && node.type.typeArguments) {
+    payloadTypes = ts.factory.createNodeArray(node.type.typeArguments.map(arg => arg as ts.TypeNode));
+  }
```
```suggestion
  if (!payloadTypes && node.type && ts.isTypeReferenceNode(node.type) && node.type.typeArguments) {
    payloadTypes = ts.factory.createNodeArray(node.type.typeArguments);
  }
```

That `map` isn't necessary

JeanMeche commented on packages/core/schematics/test/output_migration_spec.ts:
```diff
@@ -69,4 +69,46 @@ describe('output migration', () => {
+    writeFile(
+      '/foo.ts',
+      [
+        `import { Component, EventEmitter, OnInit, Output } from '@angular/core';`,
+        ``,
+        `@Directive()`,
```
```suggestion
        `import { Directive, EventEmitter, OnInit, Output } from '@angular/core';`,
        ``,
        `@Directive()`,
```

JeanMeche commented on packages/core/schematics/test/output_migration_spec.ts:
```diff
@@ -69,4 +69,46 @@ describe('output migration', () => {
+    expect(content).toContain('readonly eventMovement = output<IResponse>();');
+  });
+
+  it('should keep type also without initializer and with alias', async () => {
+    writeFile(
+      '/foo.ts',
+      [
+        `import { Component, EventEmitter, OnInit, Output } from '@angular/core';`,
+        ``,
+        `@Component({`,
+          `template: 'Hi foo'`,
+        `})`,
+        `class MyDir {`,
+        `  constructor() {}`,
+        `@Output('customEvent') eventMovement: EventEmitter<IResponse> = new EventEmitter();`,
+        `}`,
+      ].join('\n'),
+    );
```
```suggestion
  it('should keep type also without initializer and with alias', async () => {
    writeFile(
      '/foo.ts',
      [
        `import { Directive, EventEmitter, OnInit, Output } from '@angular/core';`,
        ``,
        `@Directive({})`,
        `class MyDir {`,
        `  constructor() {}`,
        `@Output('customEvent') eventMovement: EventEmitter<IResponse> = new EventEmitter();`,
        `}`,
      ].join('\n'),
    );
```
Let's keep the test a simple a possible.

JeanMeche reviewed (commented):
Looks like we're going in the right direction with this. 

Can we also add a test that covers the migration of  `@Output() eventMovement = new EventEmitter<IResponse>();` ?

JeanMeche commented on packages/core/schematics/test/output_migration_spec.ts:
```diff
@@ -69,4 +69,46 @@ describe('output migration', () => {
+    writeFile(
+      '/foo.ts',
+      [
+        `import { Component, EventEmitter, OnInit, Output } from '@angular/core';`,
+        ``,
+        `@Directive()`,
+        `class MyDir {`,
+        `  constructor() {}`,
+        `@Output() eventMovement: EventEmitter<IResponse> = new EventEmitter();`,
+        `}`,
```
```suggestion
        `import { Directive, EventEmitter, OnInit, Output } from '@angular/core';`,
        ``,
        `@Directive()`,
        `class MyDir {`,
        `@Output() eventMovement: EventEmitter<IResponse> = new EventEmitter();`,
        `}`,
```

JeanMeche commented:
Also please look at the linting failures, we have a formatting issue. 

Note: The currently broken tests are flakes and related to this change.

aparzi commented:
> Looks like we're going in the right direction with this.
> 
> Can we also add a test that covers the migration of `@Output() eventMovement = new EventEmitter<IResponse>();` ?

Sure

aparzi commented:
> Also please look at the linting failures, we have a formatting issue.
> 
> Note: The currently broken tests are flakes and related to this change.



> Also please look at the linting failures, we have a formatting issue.
> 
> Note: The currently broken tests are flakes and related to this change.

fix lint errors

JeanMeche commented:
We're still missing the test to cover `@Output() eventMovement = new EventEmitter<IResponse>();`

aparzi commented:
> We're still missing the test to cover `@Output() eventMovement = new EventEmitter<IResponse>();`

I added this test:
`@Output() eventMovement: EventEmitter = new EventEmitter<IResponse>();`

do I change it?

JeanMeche commented:
My bad, it's an oversight :)

JeanMeche reviewed (approved):
LGTM. 

However, this is outside of my review scope. I'll let Pawel do a 2nd review.

aparzi commented:
I checked inputs migration command and seems that behaviour it'ok. That command doesn't have the same bug that the output did.

pkozlowski-opensource commented on packages/core/schematics/test/output_migration_spec.ts:
```diff
@@ -69,4 +69,63 @@ describe('output migration', () => {
     const content = tree.readContent('/index.ts').replace(/\s+/g, ' ');
     expect(content).toContain('readonly out = output<string>();');
   });
+
```
This test here is just a smoke test. The actual tests verifying migration logic are in packages/core/schematics/migrations/output-migration/output-migration.spec.ts - could you please move the following tests there? Probably somewhere around the [existing test related to output declaration](https://github.com/angular/angular/blob/cbd6ec8c6a3c632089cc8668d621aa597627af16/packages/core/schematics/migrations/output-migration/output-migration.spec.ts#L29)

aparzi replied:
thanks for the tip

pkozlowski-opensource commented on packages/core/schematics/test/output_migration_spec.ts:
```diff
@@ -69,4 +69,63 @@ describe('output migration', () => {
     expect(content).toContain('readonly out = output<string>();');
   });
+
+  it('should keep type also without initializer', async () => {
```
nit: the "also" word doesn't add anything to the test description, remove?

pkozlowski-opensource reviewed (changes requested):
Thnx so much for the bug fix, appreciate it ❤️ 

The logic LGTM, just want to ask you to move tests around.

aparzi commented:
> Thnx so much for the bug fix, appreciate it ❤️
> 
> The logic LGTM, just want to ask you to move tests around.

@pkozlowski-opensource I have updated with what was requested. Thanks for your availability

For the rest it's a pleasure

aparzi committed:
fix(core): fix ng generate @angular/core:output-migration

output-migration command not keep type if @output declaring without initializer
